core/node/base_token_adjuster/src/base_token_ratio_persister.rs

```diff
@@ -124,6 +133,21 @@ impl BaseTokenRatioPersister {
             return Ok(());
         };
 
+        if let Some(ref prev_ratio) = *self.last_persisted_l1_ratio.read().unwrap() {
+            let current_ratio = BigDecimal::from(new_ratio.numerator.get())
+                .div(BigDecimal::from(new_ratio.denominator.get()));
+            let deviation = (prev_ratio - current_ratio.clone())
+                .abs()
+                .div(prev_ratio.clone())
+                .mul(BigDecimal::from(100))
+                .to_u32()
+                .unwrap();
```

Comment:
1. This will panic easily (e.g. if prev_ratio = 1/2, current_ratio = 1/3: |1/2-1/3|/(1/2)=1/3, 100/3 is not an integer). Instead convert `self.config.l1_update_deviation_percentage` to BigDecimal and compare
2. Let's extract this code in a separate function and cover with unit tests

Reply:
Can you help me to understand why it'd fail? 

is that the scenario you are talking about? 

```
    let prev_ratio = BigDecimal::from(1).div(BigDecimal::from(3));
    let current_ratio = BigDecimal::from(1).div(BigDecimal::from(2));

    let deviation = (prev_ratio.clone() - current_ratio.clone())
        .abs()
        .div(prev_ratio.clone())
        .mul(BigDecimal::from(100))
        .to_u32()
        .unwrap();

    assert_eq!(deviation, 50);
```

Reply:
Moved to a function and covered with tests anyway.

Reply:
> is that the scenario you are talking about?

No, you swapped prev_ratio and current_ratio in your example. Now looks good anyway